cni/scripts/validate_state.sh

```diff
@@ -0,0 +1,96 @@
+    echo "cilium agent : $cilium_agent"
+    
+    until cat cilium_endpoints.json
+    do
+        kubectl exec -i $cilium_agent -n kube-system -- bash -c "cilium endpoint list -o json" > cilium_endpoints.json
+        sleep 10
+    done
+
+    total_pods=$(kubectl get pods --all-namespaces -o wide --field-selector spec.nodeName=$node_name,status.phase=Running --output json)
+
+    echo "Checking if there are any pods with no ips"
+    pods_with_no_ip=$(echo $total_pods | jq -j '(.items[] | select(.status.podIP == "" or .status.podIP == null))')
+    if [ "$pods_with_no_ip" != "" ]; then
+        echo "There are some pods with no ip assigned."
+        kubectl get pods -A
+        exit 1
+    fi
+    total_pods_ips=$(echo $total_pods | jq -r '(.items[] | .status.podIP)')
+    pod_ips=()
+    num_of_pod_ips=0
+    for ip in $total_pods_ips
+    do
+        if [ $ip != $node_ip ]; then         
+            pod_ips+=($ip)
+            num_of_pod_ips=$((num_of_pod_ips+1))
+        fi
+    done
+    echo "Number of pods running with ip assigned $num_of_pod_ips"
+
+    num_of_azure_endpoint_ips=$( cat azure_endpoints.json | jq -r '[.Endpoints | .[] | .IfnameToIPMap.eth0.IPv4[0].IP] | length' )
+    azure_endpoint_ips=$( cat azure_endpoints.json | jq -r '(.Endpoints | .[] | .IfnameToIPMap.eth0.IPv4[0].IP) ' )
+    echo "Number of azure endpoint ips : $num_of_azure_endpoint_ips"
+
+    if ((num_of_pod_ips != num_of_azure_endpoint_ips)); then
+        printf "Error: Number of pods in running state is less than total ips in the azure ednpoint file" >&2 
+        exit 1
+    fi
+    echo "Azure $azure_endpoint_ips" 
+    echo "checking the ips in the azure endpoints file"
+    for ip in "${pod_ips[@]}"
+    do
+        find_in_array "$azure_endpoint_ips" "$ip" "azure_endpoints.json"
+        if [[ $? -eq 1 ]]; then
+            printf "Error: %s Not found in the azure_endpoints.json" "$ip" >&2
+            exit 1
+        fi
+    done
+
+    num_of_cilium_endpoints=$(cat cilium_endpoints.json | jq -r '[.[] | select(.status.networking.addressing[0].ipv4 != null)] | length')
+    cilium_endpoint_ips=$(cat cilium_endpoints.json | jq -r '(.[] | select(.status.networking.addressing[0].ipv4 != null) | .status.networking.addressing[0].ipv4)')
+    echo "Number of cilium endpoints: $num_of_cilium_endpoints"
+
+    if ((num_of_pod_ips != num_of_cilium_endpoints)); then
+        printf "Error: Number of pods in running state is less than total ips in the cilium endpoint file" >&2 
+        exit 1
+    fi
+
+    for ip in "${pod_ips[@]}"
+    do
+        find_in_array "$cilium_endpoint_ips" "$ip" "cilium_endpoints.json"
+        if [[ $? -eq 1 ]]; then
+            printf "Error: %s Not found in the cilium_endpoints.json" "$ip" >&2
+            exit 1
+        fi
+    done
+
+    kubectl exec -i $privileged_pod -n kube-system -- bash -c "chroot /host /bin/bash -c 'systemctl restart systemd-networkd'"
```

Comment:
Maybe add a comment here explaining (1) why we're restarting systemd-networkd (link to Cilium issue https://github.com/cilium/cilium/issues/18706 ) and (2) that there's a later step in the load test pipeline that runs a connectivity test

Reply:
Make sense, sure will add it. Thanks!